File: go/chapter4/Dlinkedlist/app/dlinkedlist/dlinkedlist.go
package dlinkedlist

type Node struct {
	data int
	next *Node
}

type List struct {
	head      *Node
	before    *Node
	cur       *Node
	numOfData int
	comp      func(d1, d2 int) int
}

func (plist *List) ListInit() {
	plist.head = &Node{next: nil}
	plist.numOfData = 0
	plist.comp = nil
}

func (plist *List) FInsert(data int) {
	newNode := &Node{data: data}

	newNode.next = plist.head.next
	plist.head.next = newNode
	plist.numOfData++
}

func (plist *List) SInsert(data int) {
	pred := plist.head
	newNode := &Node{data: data}

	for pred.next != nil && plist.comp(data, pred.next.data) != 0 {
		pred = pred.next
	}

	newNode.next = pred.next
	pred.next = newNode
	plist.numOfData++
}

func (plist *List) LInsert(data int) {
	if plist.comp == nil {
		plist.FInsert(data)
	} else {
		plist.SInsert(data)
	}
}

func (plist *List) LFirst(pdata *int) bool {
	if plist.head.next == nil {
		return false
	}

	plist.before = plist.head
	plist.cur = plist.head.next
	*pdata = plist.cur.data

	return true
}

func (plist *List) LNext(pdata *int) bool {
	if plist.cur.next == nil {
		return false
	}

	plist.before = plist.cur
	plist.cur = plist.cur.next
	*pdata = plist.cur.data

	return true
}

func (plist *List) LRemove() int {
	rdata := plist.cur.data

	plist.before.next = plist.cur.next
	plist.cur = plist.before
	plist.numOfData--

	return rdata
}

func (plist *List) LCount() int {
	return plist.numOfData
}

func SetSortRule(plist *List, comp func(d1, d2 int) int) {
	plist.comp = comp
}

File: python/chapter4/dlinkedlist/app/dlinkedlist/dlinkedlist.py
from pydantic import BaseModel, Field
from typing import Optional, Callable

class Node(BaseModel):
    data: int
    next: Optional['Node'] = None


class DLinkedList(BaseModel):
    head: Optional[Node] = Field(default_factory=lambda: Node(data=None, next=None))
    before: Optional[Node] = None
    cur: Optional[Node] = None
    num_of_data: int = 0
    comp: Optional[Callable[[int, int], int]] = None

    def FInsert(self, data: int):
        new_node = Node(data=data)
        new_node.next = self.head.next
        self.head.next = new_node
        self.num_of_data += 1

    def SInsert(self, data: int):
        pred = self.head
        new_node = Node(data=data)

        while pred.next is not None and self.comp(data, pred.next.data) != 0:
            pred = pred.next

        new_node.next = pred.next
        pred.next = new_node
        self.num_of_data+=1


    def LInsert(self, data):
        if self.comp is None:
            self.FInsert(data)
        else:
            self.SInsert(data)

    def LFirst(self) -> bool:
        if self.head.next is None:
            return False
        
        self.before = self.head
        self.cur = self.head.next
        return True

    def LNext(self) -> bool:
        if self.cur.next is None:
            return False
        
        self.before = self.cur
        self.cur = self.cur.next
        return True

    def LRemove(self):
        self.before.next = self.cur.next
        self.cur = self.before
        self.num_of_data -= 1


    def LCount(self):
        return self.num_of_data

    def SetSortRule(self, comp):
        self.comp = comp
    